esp/services/ws_workunits/ws_workunitsService.cpp

```diff
@@ -5014,3 +5014,442 @@ bool CWsWorkunitsEx::onWUGetArchiveFile(IEspContext &context, IEspWUGetArchiveFi
+
+    readQueryFileCopyErrors(publishResponse->getFileErrors(), finalResult);
+    addEclDefinitionActionResult(eclDefinition, finalResult.str(), "Publish", true, results);
+}
+
+void CWsWorkunitsEx::unpublishEclDefinition(IEspContext &context, const char *target, IEspWUEclDefinitionActionRequest &req,
+    IArrayOf<IConstWUEclDefinitionActionResult> &results)
+{
+    Owned<CWUQuerySetQueryActionRequest> actionReq = new CWUQuerySetQueryActionRequest("WsWorkunits");
+    actionReq->setQuerySetName(target);
+    actionReq->setAction("Delete");
+
+    IArrayOf<IEspQuerySetQueryActionItem> queries;
+    StringArray &eclDefinitions = req.getEclDefinitions();
+    for (aindex_t i = 0; i < eclDefinitions.length(); i++)
+    {
+        StringBuffer eclDefinitionName = eclDefinitions.item(i);
+        if (eclDefinitionName.trim().isEmpty())
+        {
+            addEclDefinitionActionResult(nullptr, "Empty ECL Definition name in Unpublish request", "Unpublish", true, results);
+            continue;
+        }
+        Owned<IEspQuerySetQueryActionItem> item = createQuerySetQueryActionItem();
+        item->setQueryId(eclDefinitionName.str());
+        queries.append(*item.getClear());
+    }
+    actionReq->setQueries(queries);
+
+    Owned<CWUQuerySetQueryActionResponse> actionResponse = new CWUQuerySetQueryActionResponse("WsWorkunits");
+    onWUQuerysetQueryAction(context, *actionReq, *actionResponse);
+
+    const IMultiException &me = actionResponse->getExceptions();
+    if (me.ordinality())
+    {
+        StringBuffer errMsg;
+        addEclDefinitionActionResult(nullptr, readExceptionMessage(me, errMsg), "Unpublish", true, results);
+        return;
+    }
+
+    IArrayOf<IConstQuerySetQueryActionResult> &actionResults = actionResponse->getResults();
+    if (actionResults.empty())
+        addEclDefinitionActionResult(nullptr, "Empty Result!", "Unpublish", true, results);
+    else
+    {
+        StringBuffer outMsg;
+        ForEachItemIn(i, actionResults)
+        {
+            IConstQuerySetQueryActionResult &item = actionResults.item(i);
+            const char *id = item.getQueryId();
+            if (item.getSuccess())
+                outMsg.appendf(" Query %s/%s success;", target, id ? id : "");
+            else if (item.getCode()|| item.getMessage())
+            {
+                const char *msg = item.getMessage();
+                outMsg.appendf(" Query %s/%s Error (%d) %s;", target, id ? id : "", item.getCode(), msg ? msg : "");
+            }
+            else
+                outMsg.appendf(" Query %s/%s failed;", target, id ? id : "");
+        }
+        addEclDefinitionActionResult(nullptr, outMsg.str(), "Unpublish", true, results);
+    }
+}
+
+bool CWsWorkunitsEx::onWUEclDefinitionAction(IEspContext &context, IEspWUEclDefinitionActionRequest &req, IEspWUEclDefinitionActionResponse &resp)
+{
+    try
+    {
+        if (!context.validateFeatureAccess(OWN_WU_ACCESS, SecAccess_Write, false))
```

Comment:
Maybe add the action type to the exception string.  It would be nice to know what the person who didn't have permission was actually trying to do.  I assume access rights to the actual repository / definition happens elsewhere?